### DISCUSSION TITLE
articles/unsafe-read-beyond-of-death/

### DISCUSSION THREAD

**Discussion by giscus** [ogxd/ogxd.github.io#12] [Blog Comments]
# articles/unsafe-read-beyond-of-death/

While developing GxHash over my spare time during my last parent leave (babies sleep a lot, I swear I’m a good father 😇), I went down the rabbit hole and came up with increasingly low-level optimizations. In this article, I want to talk about one of the nastiest ones, but to whom GxHash is owed to be the fastest non-cryptographic hashing algorithm up to this date.
I called this optimization 🥁🥁🥁 “Unsafe Read Beyond of Death” (or URBD).

https://ogxd.github.io/articles/unsafe-read-beyond-of-death/

**Comment by andersk** [comment]
This trick is considered **undefined behavior** in the Rust and LLVM memory model, even if it’s allowed by the underlying hardware. Like any undefined behavior, compilers are allowed to assume it doesn’t happen for the purpose of optimization, leading to results you don’t expect. Yes, even if you checked that you didn’t cross a page boundary—that may satisfy the hardware’s expectations, but not the compiler’s. The only way around this is to use inline assembly.

https://github.com/ogxd/gxhash/issues/82

**Reply by ogxd** [reply]
Indeed, Rust may not have been the best language for such trick. I'll add a few lines about this UB here and in the repository, and investigate what are the other options. Inlined assembly is one but back when I tried using it in 2023 it was not possible to call many of the SIMD instructions used here.

**Reply by gskt17** [reply]
I wonder if the Mill guys are willing to lease Intel their HW patents that allow this kind of operation to be safe and easy...

**Reply by andersk** [reply]
@gskt17 Again, this is [not a hardware limitation](https://www.ralfj.de/blog/2019/07/14/uninit.html); it is a [safety invariant](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers) on `unsafe` Rust pointer accesses that Rust programs must uphold in order to guarantee the correctness of the compiler’s optimizations.

**Reply by 2stein57** [reply]
I don't get why somebody would use rust for something like this anyway, this is a prime example for "unsafe" C code. I don't see the point in having discussions on what rust thinks is undefined behaviour, this is purely theoretical at this point and has nothing to do with the underlying hardware.

**Reply by andersk** [reply]
@2stein57 This is off topic, but C has these restrictions too. From the [C23 standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3220.pdf), Annex J.2 “Undefined behavior”:

> The behavior is undefined in the following circumstances:
> …
> (36) An object has its stored value accessed other than by an lvalue of an allowable type (6.5.1).
> …
> (43) Addition or subtraction of a pointer into, or just beyond, an array object and an integer type
produces a result that does not point into, or just beyond, the same array object (6.5.6).
> (44) Addition or subtraction of a pointer into, or just beyond, an array object and an integer type
produces a result that points just beyond the array object and is used as the operand of a unary
`*` operator that is evaluated (6.5.6).
> …
> (108) The pointer passed to a library function array parameter does not have a value such that all
address computations and object accesses are valid (7.1.4)
> …
> (196) A string or wide string utility function is instructed to access an array beyond the end of an
object (7.26.1, 7.31.4).

The fact that this has “nothing to do with the underlying hardware” does not mean it’s “purely theoretical”. The whole point of the [post I just linked](https://www.ralfj.de/blog/2019/07/14/uninit.html) is that C and Rust compilers can and do optimize under the assumption that your program invokes no undefined behavior **according to the language’s abstract machine**, before it ever reaches the underlying hardware. This is why the “same operation” that’s sound to write in inline assembly is unsound to write in C and Rust.

**Reply by 2stein57** [reply]
@andersk Fact is and stays: if you work on raw pointers, at least the C compiler has no knowledge about any object or sizes and just will do what you tell it to do. So my opinion: don't make it more complicated than it actually is. Who even reads these standards? I don't have time for that ;-)

**Reply by andersk** [reply]
@2stein57 Consider the function below, which takes a raw pointer and checks whether any of the first 4 bytes it points to are equal to a provided `char`. Supposing the compiler works as you describe, if we only pass a 3-byte input `"hi!"`, the function will examine whatever the following byte in memory is, and it might happen to equal `'a'` or `'b'` by chance, but it won’t be both—so the output should be one of `1 0`, `0 1`, or `0 0`:

```c
#include <stdio.h>

int contains(const char *s, char c) {
  for (int i = 0; i < 4; i++)
    if (s[i] == c)
      return 1;
  return 0;
}

int main() {
  const char s[3] = "hi!";
  printf("%d %d\n", contains(s, 'a'), contains(s, 'b'));
  return 0;
}
```

Well, if you actually [run this program](https://ato.pxeger.com/run?1=XVA5DsIwEOz9isEIxSaAuAqkJHyBBwSKkBBiCRxkOxXiJTQ0iDfxE0oWggClmtlrZnfP13SX6O0t5v3FmK8u18rl_dn90VY63VXZBqF1mSoHxZwxpR3SUrtEaSuIWAqLxKBrezVJJY4MyEsD8WpWiDAMCEJMCXxfUhVQOYSN1QpRRCPvFGA2rjIao4B9-TBgp9p1T5aiFv_ztfGENMAL1eKvsYOh3lzwToZOttS899uWNvQSTzYya0_Khl99_-cNlw8-AQ) with `clang -O2`, the output you get is something like

```
-296556408 -296556392
```

which are not even possible return values for `contains`.

(Here’s an [`unsafe` Rust equivalent](https://ato.pxeger.com/run?1=hVBLTsMwFBRbn2LqBbFRYhUKUtQoXIEDVBVygi0sFafyZxX5JGy6AA7FaXBJlLJjFu_pjeaN5r33Dxd9-NrR6onuT58x6Kr-vnrQFv1ggzTWM7_FTR58QKxL9NvcOKpHxM0dRoIMPTgYGIu1EPczd4bRiNZLrTDCi0FrrwIzXDglXxhHQtui_6M_w6kQncVts7CJXOqaJEJyuLecLDtMqwcVkENe72LdYLNHi46-mhWdLI7O2HCwK7YY0jFhTLRciMupQvrnY3CMl-gKWfB_NV3BfyW8IWn63vzE09x_AA) that outputs `1 1`, which isn’t quite as wild but is still unexpected. And I have no reason to think even wilder results aren’t possible; this is just how this particular translation happened to behave with this particular compiler.)

I hope I’ve inspired you to at least read [something about how undefined behavior actually works](https://blog.llvm.org/2011/05/what-every-c-programmer-should-know.html).

**Reply by 2stein57** [reply]
@andersk I don't per se contradict what you are trying to tell me, and quite know what compilers do and don't. And yes, I'm using them as "smart assembler" because even though I can write the stuff in assembler, I'm just 10 times faster with some help of a C++ compiler and possibly some SIMD intrinsics.
So about your example: The function you wrote with pointer is doing exactly what I would expect. 
See assembler output here: https://godbolt.org/z/9f1qKxT81
--> Clang cannot optimize it, as it has to expect that it will be called from somewhere else as well. 

In your example clang simply uses its "global" knowledge you gave it (having both caller/callee in the same compile unit), inlines the function and optimizes it away (the values you see in your printf are just old register values or stack values, depending on the actual calling convention in use).
Fun fact: gcc is not quite as agressive as clang, it also inlines and doesn't call the actual function, but is at least nice enough to actually set the input parameters of the printf function and then outputs "0 0".
I have to admit, in my last post I was lazy to mention: you should take care the compiler cannot optimize stuff away by having global knowledge (e.g. inlining or by being in the same compile unit). 
I'm sure you will now answer me that this is exactly the point you want to make, but my point is: by being careful, you definitely can use a C compiler as the tool I'm pointing out, and it is perfectly fine to do so in my opinion (and I'm doing that quite successfully for about 20 years now). 
I hope we both agree, the people that will write code that reads beyond an "array" just for SIMD optimization are perfectly capable and clever enough to see the issues and work around them, even if the "language model" was not meant to do so. And these people also will have a look at the assembler output produced, just to see what "funny stuff" the compiler is generating.

Why I'm posting all this: I just cannot agree with your first statement "The only way around this is to use inline assembly."

**Reply by andersk** [reply]
@2stein57 If you are compiling non-compliant code in the controlled environment of your own system, carefully verifying the assembly output by hand, and shipping that output as inline assembly, then that isn’t a way *around* using inline assembly, but it’s a fine way *to* use inline assembly.

On the other hand, if you are shipping C or Rust code to users, to be compiled on their system with their choice of compiler, compiler version, compiler flags, using optimizations that might not even have been invented yet, perhaps as a library to be linked with their own client code using their choice of link-time optimization settings, etc.—which is the [real context of this discussion](https://github.com/ogxd/gxhash)—then the onus is on you to comply with the standard of the language you are writing in, because you’re in no position to manually verify everything that can and will go wrong in some of these environments if you don’t.

If this feels patronizing because you already know this or think you know better, then please at least try to put yourself in the shoes of a less experienced developer reading this discussion, and consider what message they would take away from your comments as you wrote them.
